File: android/app/src/test/java/id/quezacolt/veilkeeper/crypto/AesGcmTest.kt
package id.quezacolt.veilkeeper.crypto

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertThrows
import org.junit.Test
import java.security.GeneralSecurityException
import java.security.SecureRandom

/**
 * SPEC-BASE.md Section 47 ("Critical Security Tests" -- Encryption):
 * plaintext -> encrypt -> decrypt must exactly match the original, and the
 * nonce/IV must be unique per encryption.
 */
class AesGcmTest {

    private fun randomKey() = ByteArray(AesGcm.KEY_LENGTH_BYTES).also(SecureRandom()::nextBytes)

    @Test
    fun `round trip returns the original plaintext exactly`() {
        val key = randomKey()
        val plaintext = "the quick brown fox jumps over the lazy dog".toByteArray()

        val ciphertext = AesGcm.encrypt(key, plaintext)
        val decrypted = AesGcm.decrypt(key, ciphertext)

        assertArrayEquals(plaintext, decrypted)
    }

    @Test
    fun `round trip works for empty plaintext`() {
        val key = randomKey()
        val ciphertext = AesGcm.encrypt(key, ByteArray(0))
        val decrypted = AesGcm.decrypt(key, ciphertext)
        assertEquals(0, decrypted.size)
    }

    @Test
    fun `each encryption uses a unique nonce`() {
        val key = randomKey()
        val plaintext = "same plaintext every time".toByteArray()

        val nonces = (1..50).map { AesGcm.encrypt(key, plaintext).copyOfRange(0, AesGcm.NONCE_LENGTH_BYTES) }

        val distinctNonces = nonces.map { it.toList() }.toSet()
        assertEquals("expected all 50 nonces to be unique", 50, distinctNonces.size)
    }

    @Test
    fun `same plaintext produces different ciphertext on each call`() {
        val key = randomKey()
        val plaintext = "same plaintext every time".toByteArray()

        val c1 = AesGcm.encrypt(key, plaintext)
        val c2 = AesGcm.encrypt(key, plaintext)

        assertFalse("ciphertext must differ across calls due to unique nonces", c1.contentEquals(c2))
    }

    @Test
    fun `decrypting with the wrong key fails`() {
        val ciphertext = AesGcm.encrypt(randomKey(), "secret".toByteArray())

        assertThrows(GeneralSecurityException::class.java) {
            AesGcm.decrypt(randomKey(), ciphertext)
        }
    }

    @Test
    fun `tampering with ciphertext is detected`() {
        val key = randomKey()
        val ciphertext = AesGcm.encrypt(key, "secret payload".toByteArray())
        val tampered = ciphertext.copyOf()
        tampered[tampered.size - 1] = (tampered[tampered.size - 1] + 1).toByte()

        assertThrows(GeneralSecurityException::class.java) {
            AesGcm.decrypt(key, tampered)
        }
    }

    @Test
    fun `rejects wrong key length`() {
        assertThrows(IllegalArgumentException::class.java) {
            AesGcm.encrypt(ByteArray(16), "x".toByteArray()) // AES-128 length, we require AES-256
        }
    }

    @Test
    fun `ciphertext length differs from a naive concatenation without a nonce`() {
        val key = randomKey()
        val plaintext = "x".toByteArray()
        val ciphertext = AesGcm.encrypt(key, plaintext)
        assertNotEquals(plaintext.size, ciphertext.size)
    }
}

File: web/src/crypto/__tests__/aesGcm.test.ts
import { describe, expect, it } from 'vitest'
import { encrypt, decrypt, KEY_LENGTH_BYTES, NONCE_LENGTH_BYTES } from '../aesGcm'

function randomKey(): Uint8Array {
  return crypto.getRandomValues(new Uint8Array(KEY_LENGTH_BYTES))
}

describe('aesGcm', () => {
  it('round-trips plaintext', async () => {
    const key = randomKey()
    const plaintext = new TextEncoder().encode('the quick brown fox jumps over the lazy dog')

    const ciphertext = await encrypt(key, plaintext)
    const decrypted = await decrypt(key, ciphertext)

    expect(decrypted).toEqual(plaintext)
  })

  it('round-trips with associated data', async () => {
    const key = randomKey()
    const plaintext = new TextEncoder().encode('secret content')
    const aad = new TextEncoder().encode('item-id-123')

    const ciphertext = await encrypt(key, plaintext, aad)
    const decrypted = await decrypt(key, ciphertext, aad)

    expect(decrypted).toEqual(plaintext)
  })

  it('produces a unique nonce per call (output differs each time)', async () => {
    const key = randomKey()
    const plaintext = new TextEncoder().encode('same plaintext every time')

    const a = await encrypt(key, plaintext)
    const b = await encrypt(key, plaintext)

    expect(a).not.toEqual(b)
    expect(a.slice(0, NONCE_LENGTH_BYTES)).not.toEqual(b.slice(0, NONCE_LENGTH_BYTES))
  })

  it('wire format is nonce (12 bytes) || ciphertext+tag', async () => {
    const key = randomKey()
    const plaintext = new TextEncoder().encode('x')

    const ciphertext = await encrypt(key, plaintext)

    // 12-byte nonce + 1-byte plaintext + 16-byte GCM tag
    expect(ciphertext.length).toBe(NONCE_LENGTH_BYTES + 1 + 16)
  })

  it('rejects tampered ciphertext', async () => {
    const key = randomKey()
    const plaintext = new TextEncoder().encode('do not tamper with me')

    const ciphertext = await encrypt(key, plaintext)
    const tampered = ciphertext.slice()
    tampered[tampered.length - 1] ^= 0xff

    await expect(decrypt(key, tampered)).rejects.toThrow()
  })

  it('rejects the wrong key', async () => {
    const key = randomKey()
    const wrongKey = randomKey()
    const plaintext = new TextEncoder().encode('only the right key can open this')

    const ciphertext = await encrypt(key, plaintext)

    await expect(decrypt(wrongKey, ciphertext)).rejects.toThrow()
  })

  it('rejects keys of the wrong length', async () => {
    const badKey = new Uint8Array(16) // AES-128 length, not AES-256
    const plaintext = new TextEncoder().encode('x')

    await expect(encrypt(badKey, plaintext)).rejects.toThrow()
  })
})
